oeps/oep-0041-arch-async-server-event-messaging.rst

```diff
@@ -0,0 +1,486 @@
+  infrastructure outlined here.
+
+
+----------
+Motivation
+----------
+
+Open edX has multiple services that need to copy data or take specific actions
+when certain events occur in other services. For example:
+
+* The start date of a course exists in both the Catalog domain represented by
+  the course-discovery service, as well as in the LMS, where its value is used
+  to help determine whether an enrolled student will have access to the content.
+* Other services often have a user table, with data replicated from the LMS
+  (the place where users register today).
+* Publishing a new version of course content in Studio triggers a variety of
+  LMS tasks to query and re-build optimized versions of course data for various
+  systems like the Block Transformers that power the mobile application and
+  course outline.
+* Grading events in the LMS may trigger certificate generation if the user has
+  exceeded the required threshold.
+
+Open edX services already use `Django signals
+<https://docs.djangoproject.com/en/1.11/topics/signals/>`_ for subscribing to
+events within the same service, but we have no accepted recommendation for how
+to do something similar across service boundaries. As the number of services in
+Open edX has grown, the various ad hoc synchronization and communication methods
+we've implemented over the years have caused numerous operational and
+performance problems (particularly around catalog data). At the same time,
+recent discussions around Domain Driven Design and data replication have given
+us a better framing for the kind of relationship we want to have between various
+services. There has also been a strong desire for a properly documented set of
+integration APIs that would allow Open edX to better interoperate with other
+applications in the ecosystem.
+
+There is an opportunity to bring all these elements together and create a solid
+foundation for simple, reliable, well documented, near real-time cross-service
+message passing. But without some planning and design work, we may end up in an
+even more chaotic place where each pairing of services uses different
+conventions for this kind of communication.
+
+
+-------------
+Specification
+-------------
+
+Message Format
+==============
+
+The message format will use the `CloudEvents 1.0
+<https://github.com/cloudevents/spec/blob/master/spec.md>`_ specification, with
+additional usage conventions that all Open edX services should adhere to.
+CloudEvents was developed by the Cloud Native Computing Foundation, with
+participation from major companies like Google, Microsoft, IBM, Red Hat, Oracle,
+and Huawei. While the finalized 1.0 spec is only months old and the Python
+tooling is primitive (advertised to break at any time), CloudEvents is also a
+pretty minimal message envelope spec, with the basic required fields being
+things that we'd want to have anyway.
+
+Event messages will be UTF-8 encoded JSON documents of 64 KB or less. An
```

Comment:
Reading from top to bottom (i do see a thing about callbacks but haven't read yet), but I do wonder about the 64kb limit. Not sure if that'll be enough for huge nested discovery data bits, especially if we want our messages to be stateless (And therefore contain ALL the information). probably work-around-able, but maybe something to be mindful of

Reply:
Do you mind posting a concrete example (with ID and text changed if that's necessary)? I know course-discovery has some heavily connected data models, but it's possible that we might be able to tease apart the updates into multiple events that are more reasonably sized. I think this sort of thing is likely to come up a lot, and I'd love to have some good examples here.

Reply:
Have a look at the course metadata returned by course discovery for ACCA+FA1-MA1.X -- the json metadata returned for this is 107kb

Now, this definitely is related to course runs being embedded in the metadata, and like you mention, we definitely could be judicious about how we serialize our messages... although that does bring up the stateless topic: when we're trimming metadata that is returned from a serializer, we'll need to make sure that whatever is returned is stateless in addition to being sub 64kb 

Let me know if this is enough to go on, it's just one example. Happy to dig up more!

Reply:
So I think there's an important (if sometimes blurry) distinction between stateless and denormalized. For instance, if a course instructor's `given_name` changes, I think we should send the whole staff record (`salutation`, `family_name`, etc.) I _don't_ think we should send the full data for every course run they're an instructor in.

My guess is that downstream systems that want to replicate data usually care only about a small portion of the data model, and that they're better served by stateless-but-normalized updates. So for instance, if a course run is added to a program, it's helpful for the LMS to have a message that says, "Hey, the new list of course runs for this program is [...]". But that can be a simple normalized list of IDs without all the details.

There will be exceptions of course. We're never going to echo across the list of students or the list of enrollees for a course, though it might be practical to echo the full enrollment list for an individual student. Though even then, possibly not, since that is not currently bounded.

There's also the matter of event granularity / multiple logical events. For instance, enrolling in a course and unenrolling in a course are semantically distinct and more useful as separate events than a generic, "something was updated with an Enrollment record" event. It's also probably worth specifying "enrolling myself" vs. "re-enrolling" vs. "the professor enrolled me without asking", which isn't strictly captured in that one table.

I do think there are a bunch of odd corners of this that we'll eventually have to figure out. Such as: What happens when there are related updates (e.g. a new course run is created, _and_ added to program)? Message processing is not always guaranteed to be in-order. Even if it was, it's always possible for two related messages to be distributed to two different consumers that will not see each others changes in time. In practice, I think it'll work out okay when we're trying to replicate out small bits, but even then we may have to relax certain integrity constraints...

Reply:
@macdiesel, @doctoryes: FYI in case you had thoughts on this.